### PR TITLE
feat: implement Deserializable for ProgramInfo

### PR DESCRIPTION
**Pull request by vlopes11**
Currently, `ProgramInfo` implements only `Serializable`. It is so to keep the coupling between winter-utils and miden minimal, as the serialization traits are defined in winterfell.

However, this might create some challenges for consumers of miden-vm as they might have to deserialize the struct, such as when representing the type within other targets (i.e. WASM).

This commit introduces Deserializable for Kernel and ProgramInfo.

related issue: #731

### CODE REVIEW

File: core/src/program/info.rs [hunk 1]
```diff
@@ -66,3 +69,14 @@ impl Serializable for ProgramInfo {
         <Kernel as Serializable>::write_into(&self.kernel, target);
     }
 }
+
+impl Deserializable for ProgramInfo {
```
Comment (frisitano) [inline core/src/program/info.rs:73]: Do you think we should add a test case for this?

**Review by frisitano — COMMENTED** [review]
Is it worth adding a test case for the deserialization? what do you think?

File: core/src/program/info.rs (hunk 1 above)
Comment (frisitano) [reply core/src/program/info.rs:73]: Thank you!

File: core/src/program/tests.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
+use super::{Deserializable, Digest, Felt, Kernel, ProgramInfo, Serializable};
+use crate::Word;
+use proptest::prelude::*;
+use rand_utils::prng_array;
+
+proptest! {
+    #[test]
+    fn arbitrary_program_info_serialization_works(
+        kernel_count in prop::num::u8::ANY,
+        ref seed in any::<[u8; 32]>()
```
Comment (hackaugusto) [inline core/src/program/tests.rs:10]: wouldn't this work with something like `any::<u64; 4>`? we should probably make this into a separate strategy on `crypto` later on.

**Review by frisitano — APPROVED** [review]
Looks great.  Thanks for this!

File: core/src/program/tests.rs (hunk 2 above)
Comment (vlopes11) [reply core/src/program/tests.rs:10]: This would work for a single digest. What we do there is build a stream of digests from a given seed by randomizing it at every iteration

File: core/src/program/tests.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    ) {
+        let mut seed = *seed;
+        let program_hash = digest_from_seed(seed);
+        let kernel: Vec<Digest> = (0..kernel_count)
+            .map(|_| {
+                seed = prng_array(seed);
+                digest_from_seed(seed)
```
Comment (hackaugusto) [inline core/src/program/tests.rs:17]: this is copying the same element over and over, no? maybe `vec![v; len]` is easier to read?

File: core/src/program/tests.rs (hunk 2 above)
Comment (vlopes11) [reply core/src/program/tests.rs:10]: The digests count could also be `0`, that is a very important edge case for serialization. `proptest` will definitely send a `kernel_count` `0` as it often breaks most of implementations

Comment (hackaugusto) [reply core/src/program/tests.rs:10]: humm, is that a good idea? it would throw the simplify step of the tool off, no?

File: core/src/program/tests.rs (hunk 3 above)
Comment (vlopes11) [reply core/src/program/tests.rs:17]: The idea is to make every element of the kernel distinct. If we serialize copies of `v`, the serialization implementation might just deserialize it repeatedly and the test case won't intercept such scenario.

For every call of `digest_from_seed`, we expect a different digest as we use the randomized seed as input

File: core/src/program/tests.rs (hunk 2 above)
Comment (vlopes11) [reply core/src/program/tests.rs:10]: I'm not sure wdym. One alternative is to implement [Arbitrary](https://altsysrq.github.io/rustdoc/proptest/1.0.0/proptest/prelude/trait.Arbitrary.html) for each of our types and take a vector of these.

However, we would have to think of a way of not getting very big amounts of digests. proptest would just create random size vectors, and any `usize` count of digests might be just too much for a normal desktop to handle. We do this "manual u8 digests count" because the worst case scenario, that is 255 digests, is trivial

File: core/src/program/tests.rs (hunk 3 above)
Comment (hackaugusto) [reply core/src/program/tests.rs:17]: oh, the code is overwriting the stack state on a map? I didn't notice that. Maybe use `scan` instead?

File: core/src/program/tests.rs (hunk 2 above)
Comment (vlopes11) [reply core/src/program/tests.rs:10]: They do claim however that they take the size of the type into consideration when computing the length of the vector so they will not create outrageously big data sets, but I remember having bad experiences with this in the past.

Comment (hackaugusto) [reply core/src/program/tests.rs:10]: Just to clarfiy, for this case, it probably doesn't matter, it is a simple serde test. I'm fine with merging as it is. For some more complicated proptest, then I would definitely prefer to have better shrinking without tying up the generation of different objects.

**Review by hackaugusto — APPROVED** [review]
LGTM, my comments were mostly nits which are not super important.

File: core/src/program/tests.rs (hunk 3 above)
Comment (vlopes11) [reply core/src/program/tests.rs]: Could be! As we indeed have an internal state of the iterator (the seed), it makes sense. Fixed!

File: core/src/program/tests.rs (hunk 2 above)
Comment (vlopes11) [reply core/src/program/tests.rs:10]: This is indeed a simple case. However, in general, we might indeed benefit from an engine for arbitrary sets to be used as arguments for fuzzy-tests. At least for our most atomic types (word, digest, etc).

Our general needs are simple: stream of random, uniformly distributed words. Anything we have can be derived from a word, so that should suffice.

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you!
